actionpack/lib/action_controller/metal/strong_parameters.rb

```diff
@@ -646,20 +679,114 @@ def require(key)
     #     params.permit(person: { '0': [:email], '1': [:phone]}).to_h
     #     # => {"person"=>{"0"=>{"email"=>"none@test.com"}, "1"=>{"phone"=>"555-6789"}}}
     def permit(*filters)
-      params = self.class.new
-
-      filters.flatten.each do |filter|
-        case filter
-        when Symbol, String
-          permitted_scalar_filter(params, filter)
-        when Hash
-          hash_filter(params, filter)
-        end
-      end
-
-      unpermitted_parameters!(params) if self.class.action_on_unpermitted_parameters
+      permit_filters(filters, on_unpermitted: self.class.action_on_unpermitted_parameters, explicit_arrays: false)
+    end
 
-      params.permit!
+    # `expect` is the preferred way to require and permit parameters.
+    # It is safer than the previous recommendation to call `permit` and `require`
+    # in sequence, which could allow user triggered 500 errors.
+    #
+    # `expect` is more strict with types to avoid a number of potential pitfalls
+    # that may be encountered with the `.require.permit` pattern.
+    #
+    # For example:
+    #
+    #     params = ActionController::Parameters.new(comment: { text: "hello" })
+    #     params.expect(comment: [:text])
+    #     # => #<ActionController::Parameters { text: "hello" } permitted: true>
+    #
+    #     params = ActionController::Parameters.new(comment: [{ text: "hello" }, { text: "world" }])
+    #     params.expect(comment: [:text])
+    #     # => ActionController::ParameterMissing: param is missing or the value is empty or invalid: comment
+    #
+    # In order to permit an array of parameters, the array must be defined
```

Comment:
I don't love the double brackets, but it is what it is. We could, theoretically, check if the key is pluralized and assume the values will be an array, but that probably would cause more harm than good.

Reply:
An alternative that is maybe uglier is utility method like `expect(comments: params_array(:text, :id))`. I'm not sure it's any better, but it solves the problem. `[[:text, :id]]` is the most concise way to do it that I could come up with (if only ruby had a native tuple type).

Reply:
Alternatively, it can be written as: `expect(comments: [ %i[text, id] ])` vs `expect(user: %i[name, ago])`. I avoided this format since it's requires slightly more knowledge of ruby, and is therefore slightly less approachable.